File: typescript/src/model/expressions/wdl-binary-operation.ts
/** Binary operation expression nodes and operator kinds. */
import { WdlExpressionComponentType, type WdlExpression } from './wdl-expression.js';

/** Supported binary operators from the WDL expression grammar. */
export enum WdlBinaryOperator {
  OR = '||',
  AND = '&&',
  EQ = '==',
  NEQ = '!=',
  LT = '<',
  LTE = '<=',
  GT = '>',
  GTE = '>=',
  ADD = '+',
  SUBTRACT = '-',
  MULTIPLY = '*',
  DIVIDE = '/',
  MODULO = '%',
  POWER = '**',
}

/** Models a binary expression such as `a + b` or `x && y`. */
export class WdlBinaryOperation implements WdlExpression {
  /** Creates a binary expression from its left operand, operator, and right operand. */
  public constructor(
    private leftValue?: WdlExpression,
    private operatorValue?: WdlBinaryOperator,
    private rightValue?: WdlExpression,
  ) {}

  /** Returns the left operand expression. */
  public getLeft(): WdlExpression | undefined {
    return this.leftValue;
  }
  /** Sets the left operand expression. */
  public setLeft(left: WdlExpression | undefined): void {
    this.leftValue = left;
  }
  /** Returns the operator for the expression. */
  public getOperator(): WdlBinaryOperator | undefined {
    return this.operatorValue;
  }
  /** Sets the operator for the expression. */
  public setOperator(operator: WdlBinaryOperator | undefined): void {
    this.operatorValue = operator;
  }
  /** Returns the right operand expression. */
  public getRight(): WdlExpression | undefined {
    return this.rightValue;
  }
  /** Sets the right operand expression. */
  public setRight(right: WdlExpression | undefined): void {
    this.rightValue = right;
  }

  /** Returns the broad expression family for this node. */
  public componentType(): WdlExpressionComponentType {
    return WdlExpressionComponentType.BINARY_OP;
  }
}

File: java/src/main/java/com/myriad/wdl/model/expressions/WdlBinaryOperation.java
package com.myriad.wdl.model.expressions;

import lombok.Getter;
import lombok.Setter;

public final class WdlBinaryOperation implements WdlExpression {
  public static enum Operator {
    OR("||"),
    AND("&&"),
    EQ("=="),
    NEQ("!="),
    LT("<="),
    LTE("<"),
    GT(">"),
    GTE(">="),
    ADD("+"),
    SUTRACT("-"),
    MULTIPLY("*"),
    DIVIDE("/"),
    MODULO("%"),
    POWER("**");

    @Getter private final String wdlString;

    private Operator(String wdlString) {
      this.wdlString = wdlString;
    }
  }

  @Getter @Setter private WdlExpression left;
  @Getter @Setter private Operator operator;
  @Getter @Setter private WdlExpression right;

  public WdlBinaryOperation() {}

  public WdlBinaryOperation(
      WdlExpression leftExpression, Operator operator, WdlExpression rightExpression) {
    setLeft(leftExpression);
    setOperator(operator);
    setRight(rightExpression);
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.BINARY_OP;
  }
}
